File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.py
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        counter = defaultdict(int) 
        for i in range(len(nums)):
            if i > k:
                counter[nums[i-k-1]] -= 1
            key = nums[i]
            if key in counter and counter[key] == 1:
                return True
            else:
                counter[key] = 1
        return False

File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.cpp
class Solution {
public:
    bool containsNearbyDuplicate(vector<int>& nums, int k) {
        unordered_map<int,int> freq;
        for (int i = 0; i < k && i < nums.size(); i++) {
            if (++freq[nums[i]] >= 2) return true; 
        }   
        for (int i = k; i < nums.size(); i++) {
            if (++freq[nums[i]] >= 2) return true;
            freq[nums[i-k]]--;
        }
        return false;
    }
};